clang/utils/TableGen/ClangProgModelBuiltinEmitter.cpp

```diff
@@ -848,6 +916,233 @@ void BuiltinNameEmitter::EmitQualTypeFinder() {
+                                                 : UnderlyingType::UINT;
+  } else {
+    if (T->getValueAsBit("IsFloat"))
+      ElementType = UnderlyingType::FLOAT;
+    else {
+      if (!T->isSubClassOf("FundamentalType") && !T->isSubClassOf("VectorType"))
+        ElementType = UnderlyingType::BUILTIN;
+      else
+        llvm_unreachable("Invalid type");
+    }
+  }
+  AddrSpace = StringSwitch<StringRef>(T->getValueAsString("AddrSpace"))
+                  .Case("clang::LangAS::Default", "")
+                  .Case("clang::LangAS::opencl_private", " __private")
+                  .Case("clang::LangAS::opencl_global", " __global")
+                  .Case("clang::LangAS::opencl_constant", " __constant")
+                  .Case("clang::LangAS::opencl_local", " __local")
+                  .Default(" __generic");
+}
+
+std::string JSONBuiltinInterfaceEmitter::TypeDesc::GetBaseTypeAsStr() const {
+  if (Name == "void")
+    return Name.str();
+
+  llvm::SmallString<32> Buffer;
+  {
+    llvm::raw_svector_ostream TypeStr(Buffer);
+    TypeStr << TypePrefix;
+    if (VecWidth != 1)
+      TypeStr << "_vec" << VecWidth;
+    switch (ElementType) {
+    case UnderlyingType::BOOL:
+      TypeStr << "_bool_t";
+      break;
+    case UnderlyingType::INT:
+      TypeStr << "_int" << ElementSize << "_t";
+      break;
+    case UnderlyingType::UINT:
+      TypeStr << "_uint" << ElementSize << "_t";
+      break;
+    case UnderlyingType::FLOAT:
+      TypeStr << "_fp" << ElementSize << "_t";
+      break;
+    case UnderlyingType::BUILTIN:
+      TypeStr << "_" << Name;
+      break;
+    }
+  }
+
+  return static_cast<std::string>(Buffer);
+}
+
+void JSONBuiltinInterfaceEmitter::Emit() {
+
+  // Parse the Records to populate the internal lists.
+  GetOverloads();
+  GroupBySignature();
+  ExpandTypes();
+
+  EmitBuiltins();
+}
+
+// Generate all possible types represented by the input record.
+void JSONBuiltinInterfaceEmitter::ExpandType(const Record *Ty) {
+  auto ExpansionIt = ExpandedTypes.find(Ty);
+
+  if (ExpansionIt != ExpandedTypes.end())
+    return;
+  llvm::SmallVectorImpl<std::string> &Expansion = ExpandedTypes[Ty];
+
+  JSONBuiltinInterfaceEmitter::TypeDesc TypeDesc(Ty);
+
+  bool IsCompound = Ty->isSubClassOf("CompoundType");
+  if (IsCompound) {
+    const Record *EltRecord = Ty->getValueAsDef("ElementType");
+    ExpandType(EltRecord);
+    Expansion = ExpandedTypes[EltRecord];
+    for (std::string &TypeStr : Expansion) {
+      if (TypeDesc.IsPointer) {
+        if (TypeDesc.IsConst)
+          TypeStr += " const";
+        if (TypeDesc.IsVolatile)
+          TypeStr += " volatile";
+        TypeStr += TypeDesc.AddrSpace;
+        TypeStr += " *";
+      }
+    }
+
+    return;
+  }
+
+  if (Ty->isSubClassOf("GenericType")) {
+    assert(!Expansion.size() && "Types already expanded");
+    std::vector<long int> NVec =
+        Ty->getValueAsDef("VectorList")->getValueAsListOfInts("List");
```

Comment:
Sorry for the delay.
Out internal pre-commit check failed to build this patch on Windows with the following message:
```
clang\utils\TableGen\ClangProgModelBuiltinEmitter.cpp(1026): error C2440: 'initializing': cannot convert from 'std::vector<int64_t,std::allocator<_Ty>>' to 'std::vector<long,std::allocator<_Ty>>'
         with
         [
             _Ty=int64_t
         ]
         and
         [
             _Ty=long
         ]
clang\utils\TableGen\ClangProgModelBuiltinEmitter.cpp(1026): note: No constructor could take the source type, or constructor overload resolution was ambiguous
```

Reply:
On Windows, "long" is 32 bit, so int64_t is long long.  This likely should be std::vector<int64_t> instead.